dvc/utils/threadpool.py

```diff
@@ -33,13 +33,19 @@ def create_taskset(n: int) -> Set[futures.Future]:
 
         it = zip(*iterables)
         tasks = create_taskset(self.max_workers * 5)
-        while tasks:
-            done, tasks = futures.wait(
-                tasks, return_when=futures.FIRST_COMPLETED
-            )
-            for fut in done:
-                yield fut.result()
-            tasks.update(create_taskset(len(done)))
+
+        # Yield must be hidden in closure so that the futures are submitted
+        # before the first iterator value is required.
+        def result_iterator(tasks):
```

Comment:
@skshetry had to apply a fix from the Executor.map for this, otherwise it was not working in my case

Reply:
we should probably propagate this change to dvc_obejcts/dvc_data

Reply:
@shcheklein, let's not change this. You can do something like the following:
```
with executor:
    list(executor.imap_unordered(resolve, to_resolve))
```

It is called `imap_unordered`, because it is lazy and does not guarantee ordering.

Also see [`multiprocessing.pool.Pool.imap`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.imap)
 and [`multiprocessing.pool.Pool.imap_unordered`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.imap_unordered).

Reply:
Although I am not against a new blocking `map` API.

Reply:
Okay, agreed. Let's keep it. It still stays somewhat lazy, but I see that we want it to be strictly lazy

Reply:
Let’s not change. We want lazier, stricter but one with weaker guarantees. We could add other APIs too, but that might add more maintenance burden as we also have some backports here. Also we need to keep this in sync with other projects.

It is simpler to use `list()` to make it block.